prql-elixir/native/prql/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+use rustler::{Atom, NifResult, NifStruct, NifTuple};
+
+mod atoms {
+    rustler::atoms! {
+      ok,
+      error,
+
+      // dialects
+      ansi,
+      big_query,
+      click_house,
+      generic,
+      hive,
+      mssql,
+      mysql,
+      postgres,
+      sql_lite,
+      snow_flake
+    }
+}
+
+/// Convert a `Result` from PRQL into a tuple in elixir `{:ok, binary()} | {:error, binary()}`
+fn to_result_tuple(result: Result<String, prql_compiler::ErrorMessages>) -> NifResult<Response> {
+    match result {
+        Ok(sql) => Ok(Response {
+            status: atoms::ok(),
+            result: sql,
+        }),
+        Err(e) => Ok(Response {
+            status: atoms::error(),
+            result: e.to_json(),
+        }),
+    }
+}
+
+/// Get the dialect from an atom. By default `Generic` dialect will be used
+fn dialect_from_atom(a: Atom) -> prql_compiler::sql::Dialect {
+    use prql_compiler::sql::Dialect as D;
+
+    if a == atoms::ansi() {
+        D::Ansi
+    } else if a == atoms::big_query() {
+        D::BigQuery
+    } else if a == atoms::click_house() {
+        D::ClickHouse
+    } else if a == atoms::generic() {
+        D::Generic
+    } else if a == atoms::hive() {
+        D::Hive
+    } else if a == atoms::mssql() {
+        D::MsSql
+    } else if a == atoms::mysql() {
+        D::MySql
+    } else if a == atoms::postgres() {
+        D::PostgreSql
+    } else if a == atoms::sql_lite() {
+        D::SQLite
+    } else if a == atoms::snow_flake() {
+        D::Snowflake
+    } else {
+        D::Generic
+    }
+}
+
+impl From<CompileOptions> for prql_compiler::sql::Options {
+    /// Get `prql_compiler::sql::Options` options from `CompileOptions`
+    fn from(o: CompileOptions) -> Self {
+        prql_compiler::sql::Options {
+            format: o.format,
+            dialect: Some(dialect_from_atom(o.dialect)),
+            signature_comment: o.signature_comment,
+        }
+    }
+}
+
+#[derive(Clone, NifStruct, Debug)]
+#[module = "PRQL.Native.CompileOptions"]
+pub struct CompileOptions {
```

Comment:
I guess because this needs to derive `NifStruct`, we need to copy & paste everything from upstream. IIRC we have to do the same in other bindings. 

(Would be nice to have a macro like `make_nif!(CompileOptions)` which would do the copying for us, and include upstream updates...)

Reply:
Yes, Nifs are responsible for handling the binding between Rust and BEAM VM. 
For example, in here we have to derive the `NifStruct` for a struct in Rust to be compatible with [`PRQL.Native.CompileOptions`](https://github.com/kasvith/prql/blob/9b4a17645461f13b61606b694cbc20b0e885528c/prql-elixir/lib/prql/native.ex#L14) struct in elixir.